### PR TITLE
Support cluster API objects in multiple namespaces

### LINKED ISSUE #252: Support cluster and machine in multiple namespaces in clusterctl create
Support cluster and machine in multiple namespaces in clusterctl create as it currently assumes default namespace.

### PR DESCRIPTION
**Pull request by ashish-amarnath**
**What this PR does / why we need it**:
Supporting multiple clusters requires, cluster and machine to be stored in multiple namespaces, clusterctl create currently assumes default namespace which will not work in a multi-cluster setup.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #252 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by ashish-amarnath** [comment]
@davidewatson   This is still a work in progress but just wanted to share what I have so far.

File: vendor/k8s.io/kube-openapi/example/openapi-gen/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,20 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library")
```
[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply vendor/k8s.io/kube-openapi/example/openapi-gen/BUILD.bazel]: they were made for me.

File: clusterctl/clusterdeployer/clusterdeployer.go [hunk 2]
```diff
@@ -388,6 +400,10 @@ func waitForKubeconfigReady(provider ProviderDeployer, cluster *clusterv1.Cluste
 }
 
 func pivot(from, to ClusterClient) error {
```
[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply clusterctl/clusterdeployer/clusterdeployer.go]: Yes, this is marked in my mind and will be swept after identifying all such candidates.

File: clusterctl/clusterdeployer/clusterclient.go [hunk 3]
```diff
@@ -104,10 +104,24 @@ func (c *clusterClient) Apply(manifest string) error {
...
+	return cluster, nil
+}
+
+func (c *clusterClient) GetClusterObjectsInNamespace(ns string) ([]*clusterv1.Cluster, error) {
 	clusters := []*clusterv1.Cluster{}
-	// TODO: Iterate over all namespaces where we could have Cluster API Objects https://github.com/kubernetes-sigs/cluster-api/issues/252
-	clusterlist, err := c.clientSet.ClusterV1alpha1().Clusters(apiv1.NamespaceDefault).List(metav1.ListOptions{})
+	clusterlist, err := c.clientSet.ClusterV1alpha1().Clusters(ns).List(metav1.ListOptions{})
 	if err != nil {
 		return nil, fmt.Errorf("error listing cluster objects: %v", err)
```
[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply clusterctl/clusterdeployer/clusterclient.go]: Yes, agree. Will do that.

[the remark this answers is not in the document]
Comment (ashish-amarnath) [reply clusterctl/clusterdeployer/clusterclient.go]: I took a pass for this. I will do another pass before calling this PR done.

File: clusterctl/clusterdeployer/clusterclient.go [hunk 4]
```diff
@@ -158,46 +184,52 @@ func (c *clusterClient) GetMachineObjects() ([]*clusterv1.Machine, error) {
...
 
+func (c *clusterClient) GetMachineObjects() ([]*clusterv1.Machine, error) {
+	// TODO: Iterate over all namespaces where we could have Cluster API Objects https://github.com/kubernetes-sigs/cluster-api/issues/252
+	return c.GetMachineObjectsInNamespace(apiv1.NamespaceDefault)
+}
+
 func (c *clusterClient) CreateClusterObject(cluster *clusterv1.Cluster) error {
-	// TODO: Support specific namespaces https://github.com/kubernetes-sigs/cluster-api/issues/252
-	_, err := c.clientSet.ClusterV1alpha1().Clusters(apiv1.NamespaceDefault).Create(cluster)
+	namespace := apiv1.NamespaceDefault
```
Comment (detiber) [inline clusterctl/clusterdeployer/clusterclient.go]: Instead of using apiv1.NamespaceDefault, should this rather use the namespace associated with the current kubeconfig context instead? That is the behavior I would expect, especially if passing in an existing kubeconfig for an external cluster.

Comment (ashish-amarnath) [reply clusterctl/clusterdeployer/clusterclient.go]: Good suggestion.

File: clusterctl/clusterdeployer/clusterclient.go [hunk 5]
```diff
@@ -158,46 +223,52 @@ func (c *clusterClient) GetMachineObjects() ([]*clusterv1.Machine, error) {
...
 
+func (c *clusterClient) GetMachineObjects() ([]*clusterv1.Machine, error) {
+	// TODO: Iterate over all namespaces where we could have Cluster API Objects https://github.com/kubernetes-sigs/cluster-api/issues/252
+	return c.GetMachineObjectsInNamespace(apiv1.NamespaceDefault)
+}
+
 func (c *clusterClient) CreateClusterObject(cluster *clusterv1.Cluster) error {
-	// TODO: Support specific namespaces https://github.com/kubernetes-sigs/cluster-api/issues/252
-	_, err := c.clientSet.ClusterV1alpha1().Clusters(apiv1.NamespaceDefault).Create(cluster)
+	namespace := c.configOverrides.Context.Namespace
```
Comment (ashish-amarnath) [inline clusterctl/clusterdeployer/clusterclient.go]: use `c.GetContextNamespace()` instead of `c.configOverrides.Context.Namespace`

File: clusterctl/clusterdeployer/clusterdeployer.go [hunk 6]
```diff
@@ -250,8 +277,8 @@ func (d *ClusterDeployer) createBootstrapCluster() (ClusterClient, func(), error
 	return bootstrapClient, cleanupFn, nil
 }
 
-func (d *ClusterDeployer) createTargetClusterClient(bootstrapClient ClusterClient, provider ProviderDeployer, kubeconfigOutput string) (ClusterClient, error) {
-	cluster, master, _, err := getClusterAPIObjects(bootstrapClient)
+func (d *ClusterDeployer) createTargetCluster(bootstrapClient ClusterClient, provider ProviderDeployer, kubeconfigOutput, clusterName, namespace string) (ClusterClient, error) {
```
Comment (ashish-amarnath) [inline clusterctl/clusterdeployer/clusterdeployer.go]: unintended change

File: clusterctl/clusterdeployer/clusterclient.go [hunk 7]
```diff
@@ -118,11 +171,15 @@ func (c *clusterClient) GetClusterObjects() ([]*clusterv1.Cluster, error) {
 	return clusters, nil
 }
 
-func (c *clusterClient) GetMachineDeploymentObjects() ([]*clusterv1.MachineDeployment, error) {
+func (c *clusterClient) GetClusterObjects() ([]*clusterv1.Cluster, error) {
```
Comment (ashish-amarnath) [inline clusterctl/clusterdeployer/clusterclient.go]: Preserving these methods for other users of the cluster client.
Removing these will be a breaking change.

File: clusterctl/clusterdeployer/clusterdeployer_test.go [hunk 8]
```diff
@@ -215,10 +342,11 @@ func (f *testClusterClientFactory) NewClusterClientFromKubeconfig(kubeconfig str
...
 type testProviderDeployer struct {
-	GetIPErr         error
-	GetKubeConfigErr error
-	ip               string
-	kubeconfig       string
+	GetIPErr           error
+	GetKubeConfigErr   error
+	ip                 string
+	kubeconfig         string
+	EnsureNamespaceErr error
```
Comment (ashish-amarnath) [inline clusterctl/clusterdeployer/clusterdeployer_test.go]: unnecessary change

File: clusterctl/clusterdeployer/clusterdeployer_test.go [hunk 9]
```diff
@@ -228,160 +356,279 @@ func (d *testProviderDeployer) GetKubeConfig(_ *clusterv1.Cluster, _ *clusterv1.
 	return d.kubeconfig, d.GetKubeConfigErr
 }
 
-func TestCreate(t *testing.T) {
+func (d *testProviderDeployer) EnsureNamespace(_, _ string) error {
```
Comment (ashish-amarnath) [inline clusterctl/clusterdeployer/clusterdeployer_test.go]: unused method.

File: clusterctl/clusterdeployer/clusterdeployer_test.go [hunk 10]
```diff
@@ -228,160 +356,279 @@ func (d *testProviderDeployer) GetKubeConfig(_ *clusterv1.Cluster, _ *clusterv1.
...
+			namespaceToExpectedInternalMachines: make(map[string]int),
+			namespaceToInputCluster:             map[string][]*clusterv1.Cluster{metav1.NamespaceDefault: getClustersForNamespace(metav1.NamespaceDefault, 1)},
+			cleanupExternal:                     true,
+			expectExternalCreated:               true,
+			expectErr:                           true,
 		},
 	}
+
 	for _, testcase := range testcases {
+
```
Comment (ashish-amarnath) [inline clusterctl/clusterdeployer/clusterdeployer_test.go]: unnecessary vertical whitespace

File: clusterctl/clusterdeployer/clusterclient.go (hunk 7 above)
Comment (detiber) [inline clusterctl/clusterdeployer/clusterclient.go]: Could you add comments to these preserved methods to indicate that they are deprecated and to be removed in the future? It might be good to also log the deprecation as well.

File: clusterctl/clusterdeployer/clusterdeployer.go [hunk 11]
```diff
@@ -411,92 +438,98 @@ func pivot(from, to ClusterClient) error {
...
+	}
 	if len(errors) > 0 {
 		return fmt.Errorf("error(s) encountered deleting objects from bootstrap cluster: [%v]", strings.Join(errors, ", "))
 	}
 	return nil
 }
 
-func getClusterAPIObjects(client ClusterClient) (*clusterv1.Cluster, *clusterv1.Machine, []*clusterv1.Machine, error) {
-	machines, err := client.GetMachineObjects()
+func deleteObjects(client ClusterClient) error {
```
Comment (detiber) [inline clusterctl/clusterdeployer/clusterdeployer.go]: Is this method used anywhere still?

File: clusterctl/clusterdeployer/clusterclient.go (hunk 7 above)
Comment (ashish-amarnath) [reply clusterctl/clusterdeployer/clusterclient.go]: Good suggestion. Added comments and log messages.

File: clusterctl/clusterdeployer/clusterdeployer.go (hunk 11 above)
Comment (ashish-amarnath) [reply clusterctl/clusterdeployer/clusterdeployer.go]: No. I thought I had removed it. This time really removing it :)

File: clusterctl/clusterdeployer/clusterclient.go [hunk 12]
```diff
@@ -104,12 +136,33 @@ func (c *clusterClient) Apply(manifest string) error {
...
+
+func (c *clusterClient) GetClusterObject(name, ns string) (*clusterv1.Cluster, error) {
+	clustersInNamespace, err := c.GetClusterObjectsInNamespace(ns)
+	if err != nil {
+		return nil, err
+	}
+	var cluster *clusterv1.Cluster
+	for _, nc := range clustersInNamespace {
+		if nc.Name == name {
+			cluster = nc
```
Comment (roberthbailey) [inline clusterctl/clusterdeployer/clusterclient.go]: Is there a reason not to just return here?

File: clusterctl/clusterdeployer/clusterdeployer.go [hunk 13]
```diff
@@ -111,12 +123,11 @@ const (
 	timeoutKubeconfigReady = 20 * time.Minute
 )
 
-// Creates the a cluster from the provided cluster definition and machine list.
-
+// Create the a cluster from the provided cluster definition and machine list.
```
Comment (roberthbailey) [inline clusterctl/clusterdeployer/clusterdeployer.go]: nit: create "a" or "the", not both.

File: clusterctl/clusterdeployer/clusterclient.go [hunk 14]
```diff
@@ -104,12 +136,33 @@ func (c *clusterClient) Apply(manifest string) error {
 	return c.waitForKubectlApply(manifest)
 }
 
-func (c *clusterClient) GetClusterObjects() ([]*clusterv1.Cluster, error) {
+func (c *clusterClient) GetContextNamespace() string {
+	if c.configOverrides.Context.Namespace == "" {
+		c.configOverrides.Context.Namespace = apiv1.NamespaceDefault
```
Comment (roberthbailey) [inline clusterctl/clusterdeployer/clusterclient.go]: It isn't clear to me why this needs to set the context namespace vs. just returning the default namespace. Can you explain why it's doing that?

File: clusterctl/clusterdeployer/clusterdeployer.go [hunk 15]
```diff
@@ -127,37 +138,46 @@ func (d *ClusterDeployer) Create(cluster *clusterv1.Cluster, machines []*cluster
...
 	glog.Info("Applying Cluster API stack to bootstrap cluster")
-	if err := d.applyClusterAPIStack(bootstrapClient); err != nil {
+	if err := d.applyClusterAPIStack(bootstrapClient, cluster.Namespace); err != nil {
 		return fmt.Errorf("unable to apply cluster api stack to bootstrap cluster: %v", err)
 	}
 
 	glog.Info("Provisioning target cluster via bootstrap cluster")
 
-	glog.Infof("Creating cluster object %v on bootstrap cluster", cluster.Name)
+	glog.Infof("Creating cluster object %v on bootstrap cluster in namespace %v", cluster.Name, cluster.Namespace)
```
Comment (roberthbailey) [inline clusterctl/clusterdeployer/clusterdeployer.go]: all the other places above that namespace is printed use %q instead of %v. Is there a reason that this file does it in a different way?

File: clusterctl/clusterdeployer/clusterdeployer.go [hunk 16]
```diff
@@ -411,92 +437,94 @@ func pivot(from, to ClusterClient) error {
...
 	}
-	glog.Infof("Deleting clusters")
-	if err := client.DeleteClusterObjects(); err != nil {
+	glog.Infof("Deleting clusters in namespace %q", namespace)
+	if err := client.DeleteClusterObjectsInNamespace(namespace); err != nil {
 		err = fmt.Errorf("error deleting clusters: %v", err)
 		errors = append(errors, err.Error())
 	}
+	glog.Infof("Deleting namespace %q", namespace)
+	if err := client.DeleteNamespace(namespace); err != nil {
```
Comment (roberthbailey) [inline clusterctl/clusterdeployer/clusterdeployer.go]: I believe that this will always fail for the default namespace. Have you tested this flow both with a new namespace and the default namespace?

File: vendor/k8s.io/kube-openapi/example/openapi-gen/BUILD.bazel (hunk 1 above)
[the remark this answers is not in the document]
Comment (roberthbailey) [reply vendor/k8s.io/kube-openapi/example/openapi-gen/BUILD.bazel]: Did you have to run make depend for a different part of the PR? It's usually nice to do that as a separate pull (or at least a separate commit) to make the review easier.

File: clusterctl/clusterdeployer/clusterclient.go (hunk 12 above)
Comment (ashish-amarnath) [reply clusterctl/clusterdeployer/clusterclient.go]: This was in favor of a single point of return for the method which IMO makes reading the code more easier :)

File: clusterctl/clusterdeployer/clusterdeployer.go (hunk 13 above)
Comment (ashish-amarnath) [reply clusterctl/clusterdeployer/clusterdeployer.go]: Thanks for catching that!

File: clusterctl/clusterdeployer/clusterclient.go (hunk 14 above)
Comment (ashish-amarnath) [reply clusterctl/clusterdeployer/clusterclient.go]: The intention for setting the context was to save the work of figuring out what context to use.
But I agree because the method is called `GetContextNamespace`, it should just return the default namespace.
